### PR TITLE
Rework permissions_granted to have source_group_id & origin there

### LINKED ISSUE #313: Handle ‘permissions_granted.giver_group_id’
Multiple changes/fix to be done to adapt to the latest decisions about this former `permissions_granted.giver_group_id`: ([doc](https://france-ioi.github.io/algorea-devdoc/design/access-rights/items/) updated)

* We keep this attribute, it represents through which group membership this group has received the permission. (a group (its managers) can give a permission to a single user if they want to)
* Rename `giver_group_id` to `source_group_id` (other suggestion are welcome)
* Add an `origin` attribute which defines what is the origin "process" of this perm.
  - add it to the PK
  - this is an enum:
     * group_membership: when the `group_id` is matching a non-user
     * item_unlocking: you can now easily find all perms which are probably from unlocks 
     * self: when the user set himself as owner of an item he has created (all cases where group_id=giver_group_id in former data)
     * other (to store the former way to get perms (perm given directly to the user in early days of the platform, perm given to the root items, admins), may be fixed later)
     * ...more will be added
* Set as `source_group_id` the `group_id` itself when not matching an existing group (or when 0, or -1)
* ~~To ensure this permission has been given to a group which can receive permission from this source_group_id (and ensure cascading delete if not anymore), `(source_group_id,group_id)` must refer (FK) to `groups_groups.(parent_group_id, child_group_id)` (i.e. group managers can give a permission to the whole group and a descendant of this group).~~
* ~~As a consequence, constraint `fk_permissions_granted_group_id_groups_id` on `group_id` to `groups.id` is not required anymore.~~
* Add a FK on source_group_id to `group`
* Fix `savePermissionsIntoDB`, unlocking and item creation which are not setting the right as `giver_group_id`

### PR DESCRIPTION
**Pull request by zenovich**
Fixes #313

### CODE REVIEW

**Comment by zenovich** [comment]
There are lots of weird permissions that look like given to a team or a user on contest start, but the linked items.duration = NULL. As a result, these permissions are not converted to contest participations. Instead they are marked as 'group_membership' (when group_id is a team) or 'other' (when group_id is a user).

File: db/migrations/63_get_rid_of_user_id_everywhere.sql [hunk 1]
```diff
@@ -48,9 +48,9 @@ UPDATE `groups_groups` LEFT JOIN `users` ON `users`.`id` = `groups_groups`.`invi
 ALTER TABLE `groups_groups`
     ADD CONSTRAINT `fk_groups_groups_inviting_user_id_users_group_id` FOREIGN KEY (`inviting_user_id`) REFERENCES `users`(`group_id`) ON DELETE SET NULL;
 
-ALTER TABLE `groups_items` CHANGE COLUMN `creator_user_id` `creator_id` bigint(20) DEFAULT NULL
-    COMMENT 'User who created the entry';
-UPDATE `groups_items` LEFT JOIN `users` ON `users`.`id` = `groups_items`.`creator_id` SET `groups_items`.`creator_id` = `users`.`group_id`;
+ALTER TABLE `groups_items`
+    ADD COLUMN `creator_id` bigint(20) DEFAULT NULL COMMENT 'User who created the entry' AFTER `creator_user_id` ;
```
[the remark this answers is not in the document]
Comment (zenovich) [reply db/migrations/63_get_rid_of_user_id_everywhere.sql]: fixed

File: app/api/groups/update_permissions.go [hunk 2]
```diff
@@ -74,44 +83,19 @@ func (srv *Service) updatePermissions(w http.ResponseWriter, r *http.Request) se
...
-		service.MustNotBeError(err)
-		if !found {
-			found, err = s.Items().ByID(itemID).
-				Joins("JOIN ? AS visible ON visible.item_id = items.id", itemsVisibleToGroupSubQuery).HasRows()
-			service.MustNotBeError(err)
-			if !found {
-				apiErr = service.InsufficientAccessRightsError
-				return apiErr.Error // rollback
-			}
+		apiErr = checkAccessRightsNeededToUpdatePermissions(s, user, sourceGroupID, groupID, itemID)
```
Comment (smadbe) [inline app/api/groups/update_permissions.go]: checkUserHasRightsToSetCanView vs checkAccessRightsNeededToUpdatePermissions, not clear (in their name) what rights these functions check as there are several "rights" to check.

File: app/api/groups/update_permissions.go [hunk 3]
```diff
@@ -127,6 +111,39 @@ func (srv *Service) updatePermissions(w http.ResponseWriter, r *http.Request) se
...
 
+func checkAccessRightsNeededToUpdatePermissions(s *database.DataStore, user *database.User,
+	sourceGroupID, groupID, itemID int64) service.APIError {
+	// the authorized user should be a manager of the sourceGroupID and
+	// the 'sourceGroupID' should be an ancestor of 'groupID'
+	found, err := s.Groups().ManagedBy(user).Where("groups.id = ?", sourceGroupID).
+		Joins(`
+				JOIN groups_ancestors_active AS descendants
+					ON descendants.ancestor_group_id = groups.id AND descendants.child_group_id = ?`, groupID).
+		HasRows()
```
Comment (smadbe) [inline app/api/groups/update_permissions.go]: The manager needs `can_grant_group_access` as well (on any group ancestor he is manager of)

Comment (zenovich) [reply app/api/groups/update_permissions.go]: fixed

File: app/api/groups/update_permissions.go (hunk 2 above)
Comment (zenovich) [reply app/api/groups/update_permissions.go]: fixed

**Pull request merged**